[test] Add HIP support for GPU microbenchmarks

jjotero opened the pull request:
The following tests have been modified to support both HIP and CUDA:

- kernel latency
- memory bandwidth
- shared memory bandwidth
- gpu_burn

These implementations use the Xdevice library that lives under `memory_bandwidth/src`. The other tests have symlinks to that directory.

Found a workaround to #1639 by including the symlinks as read only.

pep8speaks commented:
Hello @jjotero, Thank you for submitting the Pull Request!









* In the file [`cscs-checks/microbenchmarks/gpu/shmem/shmem.py`](https://github.com/eth-cscs/reframe/blob/97c3d89435169e9fb0a527f156376a378f816edd/cscs-checks/microbenchmarks/gpu/shmem/shmem.py):

> [Line 100:1](https://github.com/eth-cscs/reframe/blob/97c3d89435169e9fb0a527f156376a378f816edd/cscs-checks/microbenchmarks/gpu/shmem/shmem.py#L100): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

Do see the ReFrame [Coding Style Guide](https://github.com/eth-cscs/reframe/wiki/coding-style-guide)

vkarak commented on cscs-checks/microbenchmarks/gpu/memory_bandwidth/memory_bandwidth.py:
```diff
@@ -99,21 +79,57 @@ def __init__(self):
+    def select_makefile(self):
+        cp = self.current_partition.fullname
+        if cp == 'ault:amdvega':
+            self.prebuild_cmds = ['cp makefile_memoryBandwidth.hip Makefile']
```
Check my comment above.

vkarak commented on cscs-checks/microbenchmarks/gpu/memory_bandwidth/memory_bandwidth.py:
```diff
@@ -99,21 +79,57 @@ def __init__(self):
+            self.prebuild_cmds = ['cp makefile_memoryBandwidth.cuda Makefile']
+
+    @rfm.run_after('setup')
+    def set_gpu_arch(self):
```
I guess these functions will be moved to a "library test" as soon as we have have this functionality. How do you see that?

vkarak commented on cscs-checks/microbenchmarks/gpu/memory_bandwidth/memory_bandwidth.py:
```diff
@@ -13,40 +13,20 @@ def __init__(self):
                               'arolla:cn', 'tsa:cn',
                               'ault:amdv100', 'ault:intelv100',
-                              'ault:amda100']
+                              'ault:amda100', 'ault:amdvega']
```
I think that you are also missing a reference, don't you?

vkarak replied:
No, sorry, I was confused. I thought you haven't put references for all the systems, but I see they are there.

vkarak commented on cscs-checks/microbenchmarks/gpu/shmem/shmem.py:
```diff
@@ -48,3 +59,40 @@ def __init__(self):
+        # Deal with the AMD options
+        amd_trgt = None
+        if cp == 'ault:amdvega':
+            amd_trgt = 'gfx906,gfx908'
```
Question out of my ignorance: why do you need two targets?

jjotero replied:
Oh, there used to be a node in the `amdvega` partition with different AMD cards than the rest of the nodes. So that's just the product of me messing about with the target and forgetting to cleanup after Mark took that node to other partition.

jjotero commented on cscs-checks/microbenchmarks/gpu/kernel_latency/kernel_latency.py:
```diff
@@ -70,6 +48,18 @@ def __init__(self, kernel_version):
+                'ault:amda100': {
+                    'latency': (9.65, None, 0.10, 'us')
+                },
+                'ault:amdv100': {
```
Yup, that's what I meant.

vkarak commented on cscs-checks/microbenchmarks/gpu/memory_bandwidth/src/Xdevice/common/smi.hpp:
```diff
@@ -0,0 +1,117 @@
+   *   + mSize: Size (in bytes) of the total memory.
+   *
+   */
+  void getDeviceMemorySize(int id, size_t * mSize);
```
I would probably mark these as `const`. Btw, why don't you return the size and pass it as a pointer instead?

jjotero replied:
All these functions are void just to be consistent with both ROCm-SMI and NVML implementations. The functions in these libraries return a status, not the requested value. Actually, I should probably also change the getGpuTemp to be consistent with the others.

jjotero commented:
- Added the `gpu_burn` case to the PR.
-  Implemented a workaround to the symlink issue (see above) by marking the symlinks as `self.readonly_files` in the reframe test.

vkarak commented on cscs-checks/microbenchmarks/gpu/gpu_burn/src/gpu_burn.cu:
```diff
@@ -39,553 +42,369 @@
+      XMemcpy(d_B, h_B, d_resultSize, XMemcpyHostToDevice);
+  }
+
+  void compute()
```
I would only define this in the specializations and let it fail in compilation, or you could use a static assert.

jjotero replied:
I've gone for `void compute() = delete;`

vkarak replied:
Yes, better.

vkarak commented on cscs-checks/microbenchmarks/gpu/gpu_burn/src/gpu_burn.cu:
```diff
@@ -39,553 +42,369 @@
-    checkError(cuDeviceGetCount(&deviceCount));
+        // Get the time difference and return the flops
+        std::chrono::duration<double> diff = end-start;
+        double Gflops = (double)(iters*reps) * ((double)(OPS_PER_MUL)/diff.count()/1000.0/1000.0/1000.0);
```
The following, I think, should be enough. The operations will be promoted to double due to `diff.count()`.

```suggestion
        double Gflops = 1e-9 * iters * reps * OPS_PER_MUL * diff.count();
```

vkarak commented on cscs-checks/microbenchmarks/gpu/gpu_burn/src/gpu_burn.cu:
```diff
@@ -39,553 +42,369 @@
-    while (wait(NULL) != -1);
-    printf("Node %s:\n", hostname);
+    // All the burn info is stored in instances of the BurnTracker class.
+    BurnTracker * trackThreads = new BurnTracker[devCount];
```
I think you are leaking this, because you are constructing again `BurnTracker`s in the loop below with the placement new. Either make this an array of pointers to `BurnTracker`s or dummy construct them here and initialize them with an explicit method that you would call below.

jjotero replied:
Fixed. I've gone for the array of pointers approach.

vkarak commented on cscs-checks/microbenchmarks/gpu/gpu_burn/src/gpu_burn.cu:
```diff
@@ -39,553 +42,369 @@
+    // Hold until all the threads are done with the init.
+    {
+        std::unique_lock<std::mutex> lk(cv_m);
+        cv.wait(lk, []{return startUpCounter == devCount;});
```
Perhaps you can replace those with barriers in C++20.

jjotero replied:
Wouldn't this be a bit early for systems like Daint where GCC there is at 8.2?

vkarak replied:
Yes, leave it as it is now. It's just a comment for the future.

vkarak commented on cscs-checks/microbenchmarks/gpu/gpu_burn/src/gpu_burn.cu:
```diff
@@ -39,553 +42,369 @@
+    // Burn-time.
+    burn = true;
+    cv.notify_all();
+    std::this_thread::sleep_for( std::chrono::seconds(duration) );
```
What's the purpose of this sleep?

jjotero replied:
That is what sets the duration of the burn time. After that sleep, the main thread sets `burn=false` which triggers the work threads to finish up their work.

vkarak commented on cscs-checks/microbenchmarks/gpu/gpu_burn/src/gpu_burn.cu:
```diff
@@ -39,553 +42,369 @@
-        clientFaulty.push_back(false);
-    }
+    // Initialise the SMI
+    Smi * smi_handle = new Smi();
```
Couldn't you simply have `Smi smi_handle` here?

jjotero replied:
Yes - forgot to clean that up.

vkarak commented on cscs-checks/microbenchmarks/gpu/shmem/shmem.py:
```diff
@@ -48,3 +60,63 @@ def __init__(self):
         return self.job.num_tasks
+
+    @sn.sanity_function
+    def assert_num_gpus(self):
```
We should be consistent with these names. In other places we call it `assert_count_gpus()`.

jjotero commented on cscs-checks/microbenchmarks/gpu/gpu_burn/gpu_burn_test.py:
```diff
@@ -10,39 +10,24 @@
+        self.executable_opts = ['-d', '40']
+        self.build_system = 'Make'
+        self.executable = './gpu_burn.x'
+        self.readonly_files = ['Xdevice']
```
Added that to the `shmem` and `kernel_latency` tests too.

vkarak reviewed (approved):
lgtm
